File: app/src/main/java/org/movieos/feeder/fragment/EntriesFragment.java
package org.movieos.feeder.fragment;

import android.app.AlertDialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.NotNull;
import org.movieos.feeder.FeederApplication;
import org.movieos.feeder.R;
import org.movieos.feeder.databinding.EntriesFragmentBinding;
import org.movieos.feeder.databinding.EntryRowBinding;
import org.movieos.feeder.model.Entry;
import org.movieos.feeder.model.SyncState;
import org.movieos.feeder.utilities.RealmAdapter;
import org.movieos.feeder.utilities.SyncTask;

import java.text.DateFormat;

import io.realm.Realm;
import io.realm.RealmResults;
import timber.log.Timber;

public class EntriesFragment extends DataBindingFragment<EntriesFragmentBinding> {

    private static final String VIEW_TYPE = "view_type";

    RealmAdapter<Entry, EntryRowBinding> mAdapter;
    int mCurrentEntry = -1;
    private Realm mRealm;
    private int mFirstBeforePause;
    private int mLastBeforePause;

    @NonNull
    Entry.ViewType mViewType = Entry.ViewType.UNREAD;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FeederApplication.Companion.getBus().register(this);
        mRealm = Realm.getDefaultInstance();
        if (savedInstanceState != null) {
            mViewType = (Entry.ViewType) savedInstanceState.getSerializable(VIEW_TYPE);
            if (mViewType == null) {
                throw new AssertionError("null viewtype in saved instancestate");
            }
        }

        RealmResults<Entry> entries = Entry.Companion.entries(mRealm, mViewType);

        mAdapter = new RealmAdapter<Entry, EntryRowBinding>(EntryRowBinding.class, entries) {
            @Override
            public void onBindViewHolder(FeedViewHolder<EntryRowBinding> holder, Entry instance) {
                holder.getBinding().setEntry(instance);
                holder.itemView.setOnClickListener(v -> {
                    Entry.Companion.setUnread(getContext(), mRealm, Entry.Companion.byId(instance.getId()), false);
                    DetailFragment fragment = DetailFragment.create(getIds(), instance.getId());
                    fragment.setTargetFragment(EntriesFragment.this, 0);
                    getFragmentManager()
                        .beginTransaction()
                        .replace(R.id.main_content, fragment)
                        .addToBackStack(null)
                        .commit();
                });
                holder.getBinding().star.setOnClickListener(v -> {
                    boolean newState = !v.isSelected();
                    Entry.Companion.setStarred(getContext(), mRealm, instance, newState);
                    v.setSelected(newState);
                });
            }
        };
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(VIEW_TYPE, mViewType);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mRealm.close();
        FeederApplication.Companion.getBus().unregister(this);
    }


    @NonNull
    @Override
    protected EntriesFragmentBinding createBinding(@NotNull final LayoutInflater inflater, @org.jetbrains.annotations.Nullable final ViewGroup container) {
        EntriesFragmentBinding binding = EntriesFragmentBinding.inflate(inflater, container, false);
        binding.setViewType(mViewType);
        binding.recyclerView.setAdapter(mAdapter);
        binding.toolbar.inflateMenu(R.menu.entries_menu);
        binding.toolbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.menu_refresh:
                    SyncTask.Companion.sync(getActivity(), true, false);
                    return true;
                default:
                    return false;
            }
        });

        binding.stateUnread.setOnClickListener(v -> setViewType(Entry.ViewType.UNREAD));
        binding.stateStarred.setOnClickListener(v -> setViewType(Entry.ViewType.STARRED));
        binding.stateAll.setOnClickListener(v -> setViewType(Entry.ViewType.ALL));

        return binding;
    }

    @Override
    public void onResume() {
        super.onResume();
        displaySyncTime();

        // if we changed page in the detail view, scroll to minimally make that view visible.
        // To do this we tracked the first and last visible rows before we left (because in this
        // method we're not laid out yet), and will assume this has not changed. If the phone
        // has rotated or resized we'll guess wrong here.
        if (getBinding() != null && mCurrentEntry >= 0 && mFirstBeforePause >= 0 && mLastBeforePause >= 0) {
            if (mCurrentEntry < mFirstBeforePause) {
                getBinding().recyclerView.scrollToPosition(mCurrentEntry);
            } else if (mCurrentEntry > mLastBeforePause) {
                getBinding().recyclerView.scrollToPosition(mCurrentEntry - (mLastBeforePause - mFirstBeforePause));
            }
            mCurrentEntry = -1;
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (getBinding() != null) {
            LinearLayoutManager manager = (LinearLayoutManager) getBinding().recyclerView.getLayoutManager();
            mFirstBeforePause = manager.findFirstCompletelyVisibleItemPosition();
            mLastBeforePause = manager.findLastCompletelyVisibleItemPosition();
        }
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void syncStatus(SyncTask.SyncStatus status) {
        if (getBinding() == null) {
            return;
        }
        if (status.isComplete()) {
            displaySyncTime();
        } else {
            getBinding().toolbar.setSubtitle(status.getStatus());
        }
        if (status.getException() != null && isResumed()) {
            new AlertDialog.Builder(getActivity())
                .setMessage(status.getException().getLocalizedMessage())
                .setPositiveButton("OK", (dialog, which) -> dialog.dismiss())
                .show();
        }
    }

    private void displaySyncTime() {
        if (getBinding() == null) {
            return;
        }
        SyncState state = SyncState.Companion.latest(mRealm);
        DateFormat format = DateFormat.getDateTimeInstance();
        getBinding().toolbar.setSubtitle("Last synced " + (state == null ? "never" : format.format(state.getTimeStamp())));
    }


    public void childDisplayedEntryId(int entryId) {
        Timber.i("childDiplayedEntryId " + entryId);
        Entry entry = Entry.Companion.byId(entryId);
        if (entry != null && entry.isLocallyUnread()) {
            Entry.Companion.setUnread(getContext(), mRealm, entry, false);
        }
//        if (mBinding != null) {
//            mBinding.recyclerView.scrollToPosition();
//        } else {
//            mCurrentEntry = position;
//        }
    }

    private void setViewType(@NonNull Entry.ViewType viewType) {
        mViewType = viewType;
        if (getBinding() != null) {
            getBinding().setViewType(mViewType);
        }
        mAdapter.setQuery(Entry.Companion.entries(mRealm, mViewType));
    }



}

File: app/src/main/java/org/movieos/feeder/fragment/EntriesFragment.kt
package org.movieos.feeder.fragment

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.FragmentTransaction.TRANSIT_FRAGMENT_CLOSE
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import io.realm.Realm
import io.realm.RealmQuery
import io.realm.RealmResults
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.movieos.feeder.FeederApplication
import org.movieos.feeder.MainActivity
import org.movieos.feeder.R
import org.movieos.feeder.databinding.EntriesFragmentBinding
import org.movieos.feeder.databinding.EntryRowBinding
import org.movieos.feeder.model.Entry
import org.movieos.feeder.model.SyncState
import org.movieos.feeder.utilities.RealmAdapter
import org.movieos.feeder.utilities.Settings
import org.movieos.feeder.utilities.SyncTask
import timber.log.Timber
import java.text.DateFormat

class EntriesFragment : DataBindingFragment<EntriesFragmentBinding>() {

    internal var adapter: RealmAdapter<Entry, EntryRowBinding>? = null
    internal var currentEntry = -1
    private val realm: Realm = Realm.getDefaultInstance()
    private var firstBeforePause: Int = 0
    private var lastBeforePause: Int = 0

    internal var viewType: Entry.ViewType = Entry.ViewType.UNREAD
    internal var entries: RealmResults<Entry>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        FeederApplication.bus.register(this)
        if (savedInstanceState != null) {
            val viewType = savedInstanceState.getSerializable(VIEW_TYPE) as Entry.ViewType?
            this.viewType = viewType ?: this.viewType
        }

        adapter = object : RealmAdapter<Entry, EntryRowBinding>(EntryRowBinding::class.java, Entry::class.java) {
            override fun onBindViewHolder(holder: RealmAdapter.FeedViewHolder<EntryRowBinding>, instance: Entry) {
                holder.binding.entry = instance

                holder.itemView.setOnClickListener {
                    Entry.setUnread(context, realm, instance, false)
                    val fragment = DetailFragment.create(ids, instance.id)
                    fragment.setTargetFragment(this@EntriesFragment, 0)
                    fragmentManager
                            .beginTransaction()
                            .setTransition(TRANSIT_FRAGMENT_CLOSE)
                            .replace(R.id.main_content, fragment)
                            .addToBackStack(null)
                            .commit()
                }

                holder.binding.star.setOnClickListener { v ->
                    val newState = !v.isSelected
                    Entry.setStarred(context, realm, instance, newState)
                    v.isSelected = newState
                }
            }
        }

        setViewType(viewType, false)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(VIEW_TYPE, viewType)
    }

    override fun createBinding(inflater: LayoutInflater, container: ViewGroup?): EntriesFragmentBinding {
        val binding = EntriesFragmentBinding.inflate(inflater, container, false)
        binding.viewType = viewType
        binding.recyclerView.adapter = adapter
        binding.toolbar.inflateMenu(R.menu.entries_menu)
        binding.toolbar.setOnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.menu_refresh -> {
                    item.isEnabled = false
                    SyncTask.sync(activity, true, false)
                }
                R.id.menu_logout -> {
                    Settings.saveCredentials(activity, null)
                    startActivity(Intent(activity, MainActivity::class.java))
                }
            }
            true
        }

        binding.swipeRefreshLayout.setOnRefreshListener {
            binding.toolbar.menu?.findItem(R.id.menu_refresh)?.isEnabled = false
            SyncTask.sync(activity, true, false)
        }

        binding.stateUnread.setOnClickListener { setViewType(Entry.ViewType.UNREAD, true) }
        binding.stateStarred.setOnClickListener { setViewType(Entry.ViewType.STARRED, true) }
        binding.stateAll.setOnClickListener { setViewType(Entry.ViewType.ALL, true) }

        if (SyncState.latest(realm) == null) {
            // first run / first sync
            SyncTask.sync(activity, true, false)
        }

        return binding
    }

    override fun onResume() {
        super.onResume()
        displaySyncTime()

        // if we changed page in the detail view, scroll to minimally make that view visible.
        // To do this we tracked the first and last visible rows before we left (because in this
        // method we're not laid out yet), and will assume this has not changed. If the phone
        // has rotated or resized we'll guess wrong here.
        val binding = binding ?: return
        if (currentEntry >= 0 && firstBeforePause >= 0 && lastBeforePause >= 0) {
            if (currentEntry < firstBeforePause) {
                binding.recyclerView.scrollToPosition(currentEntry)
            } else if (currentEntry > lastBeforePause) {
                binding.recyclerView.scrollToPosition(currentEntry - (lastBeforePause - firstBeforePause))
            }
            currentEntry = -1
        }

        setViewType(viewType, false)
    }

    override fun onPause() {
        super.onPause()
        if (binding != null) {
            val manager = binding!!.recyclerView.layoutManager as LinearLayoutManager
            firstBeforePause = manager.findFirstCompletelyVisibleItemPosition()
            lastBeforePause = manager.findLastCompletelyVisibleItemPosition()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        realm.close()
        FeederApplication.bus.unregister(this)
    }



    @Suppress("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun syncStatus(status: SyncTask.SyncStatus) {
        if (status.isComplete) {
            setViewType(viewType, binding?.swipeRefreshLayout?.isRefreshing ?: false)
            binding?.swipeRefreshLayout?.isRefreshing = false
            displaySyncTime()
            binding?.toolbar?.menu?.findItem(R.id.menu_refresh)?.isEnabled = true
        } else {
            binding?.toolbar?.subtitle = status.status
        }
        if (status.exception != null && isResumed) {
            AlertDialog.Builder(activity)
                    .setMessage(status.exception!!.localizedMessage)
                    .setPositiveButton("OK") { dialog, _ -> dialog.dismiss() }
                    .show()
        }
    }

    private fun displaySyncTime() {
        val state = SyncState.latest(realm)
        val format = DateFormat.getDateTimeInstance()
        binding?.toolbar?.subtitle = "Last synced " + if (state == null) "never" else format.format(state.timeStamp)
    }


    fun childDisplayedEntryId(entryId: Int) {
        Timber.i("childDiplayedEntryId " + entryId)
        val entry = Entry.byId(realm, entryId).findFirst()
        if (entry != null && entry.unread && context != null) {
            Entry.setUnread(context, realm, entry, false)
        }
    }

    private fun setViewType(viewType: Entry.ViewType, clearState: Boolean) {
        // If we're keeping the same viewtype, then never _remove_ entries
        val currentIds = if (this.viewType == viewType && !clearState) entries?.map{ it.id } else null

        this.viewType = viewType
        binding?.viewType = this.viewType

        entries?.removeAllChangeListeners()
        entries = entries(realm, this.viewType, currentIds)
        entries?.addChangeListener({ results, _ ->
            adapter?.entries = results
            if (results.size == 0) {
                binding?.empty?.visibility = View.VISIBLE
            } else {
                binding?.empty?.visibility = View.GONE
            }
        })
    }

    fun entries(realm: Realm, viewType: Entry.ViewType, currently: List<Int>?): RealmResults<Entry> {
        // TODO bug, the first fime we build this it' sa raw query. So if we mark something as
        // read before we rebuild the list for any reason, that entry will dissapear.
        Timber.i("currently is $currently")
        val entries: RealmQuery<Entry> = when (viewType) {
            Entry.ViewType.UNREAD ->
                realm.where(Entry::class.java).equalTo("unread", true)
            Entry.ViewType.STARRED ->
                realm.where(Entry::class.java).equalTo("starred", true)
            Entry.ViewType.ALL ->
                realm.where(Entry::class.java)
        }
        currently?.forEach {
            entries.or().equalTo("id", it)
        }
        return entries.findAllSortedAsync("published", io.realm.Sort.DESCENDING)
    }

    companion object {
        private val VIEW_TYPE = "view_type"
    }


}